src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs

```diff
@@ -0,0 +1,415 @@
+        /// <returns>The mapped Id of the token.</returns>
+        public override int? TokenToId(string token, bool skipSpecialTokens)
+        {
+            if (string.IsNullOrEmpty(token))
+            {
+                return 0;
+            }
+
+            if (!skipSpecialTokens && _specialTokensEncoder is not null && _specialTokensEncoder.TryGetValue(token, out int specialTokenId))
+            {
+                return specialTokenId;
+            }
+
+            if (_cache.Lookup(token, out int[] ids))
+            {
+                if (ids.Length == 1)
+                {
+                    return ids[0];
+                }
+
+                return null;
+            }
+
+            if (_vocab.TryGetValue(token, out int id))
+            {
+                return id;
+            }
+
+            int[] idsToCache = BytePairEncoder.BytePairEncode(Encoding.UTF8.GetBytes(token), _encoder);
+            _cache.Add(token, idsToCache.ToArray());
+
+            if (idsToCache.Length == 1)
+            {
+                return idsToCache[0];
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Map the tokenized Id to the token.
+        /// </summary>
+        /// <param name="id">The Id to map to the token.</param>
+        /// <param name="skipSpecialTokens">Indicate if want to skip the special tokens during the decoding.</param>
+        /// <returns>The mapped token of the Id.</returns>
+        public override string? IdToToken(int id, bool skipSpecialTokens = false)
+        {
+            if (!skipSpecialTokens && _specialTokensDecoder is not null && _specialTokensDecoder.TryGetValue(id, out string token))
+            {
+                return token;
+            }
+
+            if (_decoder.TryGetValue(id, out byte[] tokenBytes))
+            {
+                return Encoding.UTF8.GetString(tokenBytes);
+            }
+
+            return null;
+        }
+
+        internal string? IdsToString(IEnumerable<int> ids, bool skipSpecialTokens = false)
+        {
+            if (ids is null)
+            {
+                return null;
+            }
+
+            List<byte> utf8Bytes = new();
+            bool useSpecialTokens = !skipSpecialTokens && _specialTokensDecoder is not null;
+
+            foreach (int id in ids)
+            {
+                if (_decoder.TryGetValue(id, out byte[] tokenBytes))
+                {
+                    utf8Bytes.AddRange(tokenBytes);
+                }
+                else if (useSpecialTokens && _specialTokensDecoder!.TryGetValue(id, out string token))
+                {
+                    utf8Bytes.AddRange(Encoding.UTF8.GetBytes(token));
+                }
+                else
+                {
+                    return null;
+                }
+            }
+
+            return utf8Bytes.Count > 0 ? Encoding.UTF8.GetString(utf8Bytes.ToArray()) : string.Empty;
```

Comment:
Do we only target netstandard2.0, or do we multitarget and build this for netcoreapp as well?  There are newer APIs that make this cheaper.

Reply:
Tracked through the issue https://github.com/dotnet/machinelearning/issues/6989